File: PitchDetection/FeedbackSystem.cpp
#include "FeedbackSystem.h"

FeedbackSystem::FeedbackSystem() {
    cents = 0.0;
    overall = 0;
    overallCount = 0;
}

void FeedbackSystem::collectData(int index, double cent) {
    if(abs(cent) <= 5){
        pitchClass[index]++;
        overall++;
    }
    pitchCount[index]++;
    overallCount++;
    cents += abs(cent);
}

double FeedbackSystem::getOverall() {
    return (100.0 * overall)/overallCount;
}

void FeedbackSystem::displayData() {
    double pitchError[12];
    for(int i = 0; i < 11; i++){
        pitchError[i] = (100.0 * pitchClass[i]) / pitchCount[i];
    }
    double avgCents = cents/overallCount;

    cout << "Here are the results of this session:" << endl;
    cout << endl;
    cout << "These are your accuracies for each pitch class:" << endl;

    for(int i = 0; i < 11; i++){
        if(isnan(pitchError[i])){
            cout << notes[i] << " was not played/sung in the session." << endl;
        }
        else{
            cout << notes[i] << " was in tune for " << pitchError[i] << "% of the time." << endl;
        }
    }
    cout << endl;
    cout << "Overall:" << endl;
    cout << "You were in tune for " << getOverall() << "% of the time." << endl;
    cout << "You were off by an average of " << avgCents << " cents." << endl;
}

File: PitchPython/FeedbackSystem.py
from queue import Queue
import math


class FeedbackSystem:
    def __init__(self, range):
        self._notes = ("C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B")
        self._pitch_class = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        self._pitch_count = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        self._cents = 0.0
        self._overall = 0
        self._overallCount = 0
        self._threshold = range
        self._recent_notes = Queue(maxsize=8)

    def get_overall(self):
        return (100.0 * self._overall) / self._overall

    def get_recent_notes(self):
        return self._recent_notes

    def _collect_data(self, index, cent):
        if abs(cent) <= self._threshold:
            self._pitch_class[index] += 1
            self._overall += 1
        self._pitch_count[index] += 1
        self._overallCount += 1
        self._cents += abs(cent)

        # May need to put an if statement for when the current note is the same as the last note
        self._recent_notes.put(self._notes[index])
        if self._recent_notes.full():
            self._recent_notes.get()



    def _display_data(self):
        avg_cents = self._cents / self._overallCount

        print("These are your accuracies for each pitch class:")
        for i in range(12):
            pitch_error = (100.0 * self._pitch_class[i]) / self._pitch_count[i]
            if math.isnan(pitch_error):
                print(self._notes[i], "was not played/sung in the session.")
            else:
                print("%s was in tune for %.4f %% of the time." % (self._notes[i], pitch_error))

        print("")
        if math.isnan(self.get_overall()):
            print("There was no audio input.")
        else:
            print("Overall:")
            print("You were in tune for %.4f %% of the time." % self.get_overall())
            print("You were off by an an average of %.4f cents." % avg_cents)
